lib/src/logger/terse.dart

```diff
@@ -0,0 +1,58 @@
+// Copyright 2018 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'package:collection/collection.dart';
+import 'package:source_span/source_span.dart';
+import 'package:stack_trace/stack_trace.dart';
+
+import '../logger.dart';
+
+/// The maximum number of repetitions of the same warning [TerseLogger] will
+/// emit before hiding the rest.
+const _maxRepetitions = 5;
+
+/// A logger that wraps an inner logger to omit repeated deprecation warnings.
+///
+/// A warning is considered "repeated" if the first paragraph is the same as
+/// another warning that's already been emitted.
+class TerseLogger implements Logger {
+  /// A map from the first paragraph of a warning to the number of times this
+  /// logger has emitted a warning with that line.
+  final _warningCounts = <String, int>{};
+
+  final Logger _inner;
+
+  TerseLogger(this._inner);
+
+  void warn(String message,
+      {FileSpan? span, Trace? trace, bool deprecation = false}) {
+    if (deprecation) {
+      var firstParagraph = message.split("\n\n").first;
```

Comment:
My only concern is that we're now assigning special semantics to `\n\n` in a warning, which could cause confusion. But, I'm not sure if any other heuristic would make sense here.

Reply:
We could theoretically try to pass in some sort of side-channel "deprecation type" identifier, maybe through Zones, but I think that would be more complicated than would be necessary. Given that we control both this heuristic *and* the messages that it applies to, I think it's fine.